// react-native-auth0/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  TimeoutError: true,
  useAuth0: true,
  Auth0Provider: true,
  LocalAuthenticationOptions: true,
  LocalAuthenticationLevel: true,
  LocalAuthenticationStrategy: true
};
Object.defineProperty(exports, "Auth0Provider", {
  enumerable: true,
  get: function () {
    return _auth0Provider.default;
  }
});
Object.defineProperty(exports, "LocalAuthenticationLevel", {
  enumerable: true,
  get: function () {
    return _localAuthenticationLevel.default;
  }
});
Object.defineProperty(exports, "LocalAuthenticationOptions", {
  enumerable: true,
  get: function () {
    return _localAuthenticationOptions.default;
  }
});
Object.defineProperty(exports, "LocalAuthenticationStrategy", {
  enumerable: true,
  get: function () {
    return _localAuthenticationStrategy.default;
  }
});
Object.defineProperty(exports, "TimeoutError", {
  enumerable: true,
  get: function () {
    return _fetchWithTimeout.TimeoutError;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "useAuth0", {
  enumerable: true,
  get: function () {
    return _useAuth.default;
  }
});
var _fetchWithTimeout = require("./utils/fetchWithTimeout");
var _useAuth = _interopRequireDefault(require("./hooks/use-auth0"));
var _auth0Provider = _interopRequireDefault(require("./hooks/auth0-provider"));
var _localAuthenticationOptions = _interopRequireDefault(require("./credentials-manager/localAuthenticationOptions"));
var _localAuthenticationLevel = _interopRequireDefault(require("./credentials-manager/localAuthenticationLevel"));
var _localAuthenticationStrategy = _interopRequireDefault(require("./credentials-manager/localAuthenticationStrategy"));
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
var _auth = _interopRequireDefault(require("./auth0"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = _auth.default;
